### PR TITLE
fix(producer): inline CDN scripts for offline render and detect black video

### PR DESCRIPTION
**Pull request by miguel-heygen**
## Context

`npx hyperframes render` fails to load assets and produces all-black video. Root causes:

1. CDN scripts (GSAP, Lottie) are left as `<script src="https://...">` in compiled HTML — the headless browser must fetch them over the network, which fails in Docker, CI, and firewalled environments
2. Assets referenced from outside the project directory (e.g. `../shared-assets/hero.png`) 404 because the file server only serves from `projectDir`
3. When GSAP fails to load, the timeline never registers, all `.clip` elements stay `visibility: hidden`, and every frame is black — with no diagnostic output
4. The linter reports `missing_gsap_script` when GSAP is bundled inline (no `<script src>` tag), which blocks users from working around the CDN issue

### 1\. CDN script inlining (`htmlCompiler.ts`)

`compileForRender` now downloads all external `<script src="https://...">` tags at compile time and inlines their content into the HTML. Rendering no longer needs network access.

| Before | After |
| --- | --- |
| `<script src="https://cdn.jsdelivr.net/npm/gsap@3.12.5/dist/gsap.min.js"></script>` stays in HTML, browser must fetch at runtime | Script downloaded during compilation, embedded as `<script>/* inlined: https://... */\n...code...</script>` |
| Docker/CI render → `net::ERR_NAME_NOT_RESOLVED` → black video | Render works fully offline |
| No feedback when CDN fails | `[Compiler] WARNING: Failed to download CDN script: ... Consider bundling it locally` |

### 2\. External asset copying (`htmlCompiler.ts`, `renderOrchestrator.ts`)

After compilation, the HTML is scanned for `src`, `href`, and CSS `url()` references that resolve outside `projectDir`. These files are copied into the compiled output directory so the file server can serve them.

| Before | After |
| --- | --- |
| `background-image: url(../shared-assets/hero.png)` → 404 (file server can't serve outside `projectDir`) | Asset detected, copied to compiled dir, path rewritten → serves correctly |
| `` → 404 | Same fix — works for all `src`/`href` attributes and CSS `url()` |

### 3\. Black video diagnostics (`renderOrchestrator.ts`)

When composition duration is 0 (which would produce a black video), the error now probes the browser for diagnostics instead of a generic message.

| Before | After |
| --- | --- |
| `Invalid composition duration: 0. Check that GSAP timelines are registered.` | `Composition duration is 0 — this would produce a black video.\n\nDiagnostics:\n  - GSAP is not loaded — CDN script may have failed to download. Bundle GSAP locally...\n  - Browser: [Browser:PAGEERROR] gsap is not defined` |
| Asset 404s during page load silently logged | `[Render] Asset load failure: ...` + `[WARN] Browser encountered network failures during page load` |

### 4\. Linter: recognize inline GSAP (`core/lint/rules/gsap.ts`)

The `missing_gsap_script` rule now recognizes GSAP bundled inline — matching the producer's inlining comment (`/* inlined: ...gsap... */`), GSAP library internals (`_gsScope`, `GreenSock`), and large inline scripts (>5KB) referencing gsap.

| Before | After |
| --- | --- |
| User inlines GSAP → linter errors with `missing_gsap_script` | Inline GSAP detected, no false error |
| Producer inlines CDN → linter errors on the compiled HTML | Producer's `/* inlined: ... */` comment recognized |

## Test plan

- [x] `pnpm build` passes
- [x] Core tests pass (410/410, +2 new)
- [x] **Reproduced baseline failures on** **`main`**: CDN scripts not inlined, external assets 404, no diagnostics
- [x] **Verified fixes**: CDN script inlined, external assets copied and served, diagnostics printed
- [x] Render with working CDN → `[Compiler] Inlined CDN script: ...` → render succeeds
- [x] Render with broken CDN → `[Compiler] WARNING: Failed to download CDN script` + browser errors surfaced
- [x] Render with assets outside project dir → `[Compiler] Found 1 asset(s) outside project directory` → assets served correctly
- [x] Linter with inline GSAP → no `missing_gsap_script` false positive

### CODE REVIEW

File: packages/producer/src/services/htmlCompiler.ts
Comment (jrusso1020) [inline packages/producer/src/services/htmlCompiler.ts:1]: **Missing unit tests for the two biggest new functions.**

`inlineExternalScripts` and `collectExternalAssets` are ~140 lines of new logic with HTML parsing, regex replacement, filesystem access, and network fetches — but neither has a unit test. `htmlCompiler.ts` has no test file at all today.

The functional tests described in the PR body are great for validation, but these functions have enough edge cases that unit tests would catch regressions and make future refactors safe:

- `inlineExternalScripts`: what happens with duplicate `src` attributes? Scripts with content between open/close tags? Multiple scripts pointing to the same CDN URL?
- `collectExternalAssets`: absolute paths in `src` (e.g. `src="/usr/share/fonts/foo.woff"`), relative paths that resolve inside `projectDir`, `data:` URIs, paths with `..` that resolve back into `projectDir`

Not a blocker for merge, but I'd strongly recommend adding an `htmlCompiler.test.ts` as a fast follow — these functions are on the critical render path.

File: packages/producer/src/services/htmlCompiler.ts [hunk 1]
```diff
@@ -692,6 +694,146 @@ function ensureFullDocument(html: string): string {
...
+  );
+
+  let result = html;
+  for (let i = 0; i < downloads.length; i++) {
+    const download = downloads[i]!;
+    const { src } = externalScripts[i]!;
+    if (download.status === "fulfilled") {
+      const escapedSrc = src.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
+      const scriptTagRe = new RegExp(
+        `<script\\b[^>]*\\bsrc=["']${escapedSrc}["'][^>]*>\\s*</script>`,
```
Comment (jrusso1020) [inline packages/producer/src/services/htmlCompiler.ts:732]: nit: This escaping is correct (`<\/script` breaks the HTML parser's closing-tag detection while JS evaluates `\/` as `/`), but it's non-obvious enough to trip up future readers. Consider a brief comment:

```ts
// Escape </script in downloaded content to prevent premature tag closure.
// <\/script is safe: HTML parser doesn't recognize it as a close tag,
// but JS treats \/ as / so the code executes identically.
```

File: packages/producer/src/services/htmlCompiler.ts [hunk 2]
```diff
@@ -692,6 +694,146 @@ function ensureFullDocument(html: string): string {
...
+  if (externalScripts.length === 0) return html;
+
+  const downloads = await Promise.allSettled(
+    externalScripts.map(async ({ src }) => {
+      const response = await fetch(src, {
+        signal: AbortSignal.timeout(15_000),
+      });
+      if (!response.ok) throw new Error(`HTTP ${response.status} for ${src}`);
+      return { src, text: await response.text() };
+    }),
```
Comment (jrusso1020) [inline packages/producer/src/services/htmlCompiler.ts:722]: The function parses the DOM to *find* scripts but then does regex replacement on the raw HTML string. This works — and I understand it's to avoid linkedom/happy-dom round-trip mangling the rest of the document — but it means the regex and the DOM query could disagree on edge cases (e.g., `src` in a comment, conditional HTML, non-standard quoting).

For now this is fine since CDN scripts are typically straightforward `<script src="..."></script>` tags. Just flagging the implicit coupling: if the DOM parse and regex ever diverge, the replacement silently becomes a no-op (regex doesn't match → original HTML returned unchanged → CDN script not inlined → black video in offline render, with only the "Failed to download" warning as a clue).

File: packages/producer/src/services/htmlCompiler.ts [hunk 3]
```diff
@@ -692,6 +694,146 @@ function ensureFullDocument(html: string): string {
...
+): { html: string; externalAssets: Map<string, string> } {
+  const absProjectDir = resolve(projectDir);
+  const externalAssets = new Map<string, string>();
+  const CSS_URL_RE = /\burl\(\s*(["']?)([^)"']+)\1\s*\)/g;
+
+  function processPath(rawPath: string): string | null {
+    const trimmed = rawPath.trim();
+    if (
+      !trimmed ||
+      trimmed.startsWith("http://") ||
```
Comment (jrusso1020) [inline packages/producer/src/services/htmlCompiler.ts:773]: Note: `resolve(absProjectDir, trimmed)` — if `trimmed` is an absolute path (e.g., `src="/etc/passwd"`), `resolve()` ignores `absProjectDir` entirely. The file then passes the "outside projectDir" check, `existsSync` returns true, and it gets copied into the compile dir.

This isn't really a security issue since this is a local build tool (the user authors the HTML and already has filesystem access), but it's worth being aware of. If you ever want to tighten this, you could skip absolute paths:

```ts
if (trimmed.startsWith("/")) return null;
```

File: packages/producer/src/services/renderOrchestrator.ts [hunk 4]
```diff
@@ -611,11 +623,58 @@ export async function executeRenderJob(
...
+      if (probeSession) {
+        const timelinesInfo = await probeSession.page.evaluate(() => {
+          const tl = (window as any).__timelines;
+          const hf = (window as any).__hf;
+          return {
+            timelineKeys: tl ? Object.keys(tl) : [],
+            hfDuration: hf?.duration ?? null,
+            gsapLoaded: typeof (window as any).gsap !== "undefined",
+          };
+        });
```
Comment (jrusso1020) [inline packages/producer/src/services/renderOrchestrator.ts]: Nice diagnostics. One thing to consider: `probeSession.page.evaluate()` can throw if the browser tab crashed or the page is in a bad state (which might be *why* duration is 0). The outer try/catch at line ~370 will catch it, but the error that surfaces would be the Puppeteer error ("Target closed" etc.) rather than the more helpful "Composition duration is 0" message.

Consider wrapping the diagnostics gathering in its own try/catch so a failed probe doesn't mask the real error:

```ts
let diagnostics: string[] = [];
try {
  if (probeSession) {
    const timelinesInfo = await probeSession.page.evaluate(() => { ... });
    // ... build diagnostics
  }
} catch {
  diagnostics.push("(Could not gather browser diagnostics — page may have crashed)");
}
```

File: packages/core/src/lint/rules/gsap.test.ts [hunk 5]
```diff
@@ -249,6 +249,44 @@ describe("GSAP rules", () => {
...
+  <div data-composition-id="main" data-width="1920" data-height="1080"></div>
+  <script>/* inlined: https://cdn.jsdelivr.net/npm/gsap@3.12.5/dist/gsap.min.js */
+    !function(t,e){t.gsap=e()}(this,function(){return {}});
+  </script>
+  <script>
+    window.__timelines = window.__timelines || {};
+    const tl = gsap.timeline({ paused: true });
+    tl.to("#box", { x: 100, duration: 1 }, 0);
+    window.__timelines["main"] = tl;
+  </script>
```
Comment (jrusso1020) [inline packages/core/src/lint/rules/gsap.test.ts:289]: Good positive test cases. Consider adding a **negative** test: a small inline script that *uses* gsap but does NOT bundle the library (no `GreenSock`, no `_gsScope`, under 5KB, no inlining comment). This should still fire `missing_gsap_script`. Without it, you can't be sure the new `hasInlineGsap` check doesn't accidentally suppress the warning for users who just write `gsap.to(...)` in an inline script without loading the library at all.

**Review by jrusso1020 — COMMENTED** [review]
## Staff Review — PR #182

**Overall: Well-crafted PR that solves a real, painful user problem. Approve with minor suggestions.**

### What's good
- **Excellent PR description** — the before/after tables make it immediately clear what changed and why. This is how all PRs should be documented.
- **Four cohesive fixes** that address the root causes rather than papering over symptoms: CDN inlining, external asset copying, black video diagnostics, and linter updates.
- **Defense-in-depth** on the asset copying path — `resolve()` canonicalization + `startsWith` check in the orchestrator prevents path traversal.
- **Graceful degradation** — if CDN download fails, it warns instead of crashing, and the diagnostics section tells the user *exactly* what went wrong.
- **`Promise.allSettled`** for CDN fetches is the right choice — one failed download shouldn't block others.

### Suggestions (none blocking)

1. **Unit tests for `inlineExternalScripts` and `collectExternalAssets`** — these are ~140 lines of new logic on the critical render path with HTML parsing, regex, filesystem access, and network fetches. `htmlCompiler.ts` has no test file today. Strongly recommend adding `htmlCompiler.test.ts` as a fast follow.

2. **Wrap diagnostics `page.evaluate()` in try/catch** — if the browser tab crashed (which could be *why* duration is 0), the Puppeteer error would mask the more helpful "duration is 0" message.

3. **Negative linter test case** — add a test confirming that a small inline script using `gsap.to()` without bundling the library still triggers `missing_gsap_script`. Ensures the new `hasInlineGsap` check doesn't accidentally suppress real warnings.

4. Minor nits on comments and edge case awareness flagged inline.

### Verdict
The code is safe, well-structured, and directly addresses the user-reported issues from the Slack thread. The missing unit tests are a gap I'd want closed soon, but not a merge blocker given the functional testing described. Ship it. 🚢

File: packages/producer/src/services/htmlCompiler.ts
Comment (miguel-heygen) [reply packages/producer/src/services/htmlCompiler.ts:1]: Agreed — these are on the critical render path and deserve proper unit tests. The functions are testable in isolation (DOM parsing + filesystem for `collectExternalAssets`, mocked fetch for `inlineExternalScripts`). Will add `htmlCompiler.test.ts` as a fast follow.

File: packages/producer/src/services/htmlCompiler.ts (hunk 1 above)
Comment (miguel-heygen) [reply packages/producer/src/services/htmlCompiler.ts:732]: Added the expanded comment in a382e51 — explains why `<\/script` is safe (HTML parser doesn't match, JS treats \\/  as /).

File: packages/producer/src/services/htmlCompiler.ts (hunk 2 above)
Comment (miguel-heygen) [reply packages/producer/src/services/htmlCompiler.ts:722]: Good flag. The DOM-then-regex approach is intentional to avoid linkedom round-trip mangling. The risk (silent no-op on mismatch) is mitigated by the `[Compiler] Inlined CDN script` log — if the count doesn't match expected scripts, it's visible. For now this covers the CDN `<script src="https://..."></script>` pattern reliably. If we hit edge cases, we can switch to full DOM manipulation.

File: packages/producer/src/services/htmlCompiler.ts (hunk 3 above)
Comment (miguel-heygen) [reply packages/producer/src/services/htmlCompiler.ts:773]: Added the `trimmed.startsWith("/")  → return null` guard in a382e51. Absolute paths in HTML src/href are rare and don't make sense to copy into the compiled dir.

File: packages/producer/src/services/renderOrchestrator.ts (hunk 4 above)
Comment (miguel-heygen) [reply packages/producer/src/services/renderOrchestrator.ts]: Wrapped the diagnostics in try/catch in a382e51. If the page crashed, it now falls back to `"(Could not gather browser diagnostics — page may have crashed)"` instead of masking the duration=0 error.

File: packages/core/src/lint/rules/gsap.test.ts (hunk 5 above)
Comment (miguel-heygen) [reply packages/core/src/lint/rules/gsap.test.ts:289]: Added in a382e51 — test confirms a small inline script using `gsap.timeline()` without bundling the library still fires `missing_gsap_script`. All 15 GSAP lint tests pass.

**Review by jrusso1020 — APPROVED** [review]

**Comment by miguel-heygen** [comment]
### Merge activity

* **Apr 1, 9:15 PM UTC**: @miguel-heygen merged this pull request with [Graphite](https://app.graphite.com/github/pr/heygen-com/hyperframes/182).
